lua_modules/mcp23017/mcp23017.lua

```diff
@@ -0,0 +1,300 @@
+    A single pin can be set or a whole byte. The numbering of the individual pins starts at 0 and ends with 15.
+    The numbers up to 7 are on register GPIO A and from 8 to 15 on GPIO B.
+
+	The module requires `i2c` and `bit` C module built into firmware.
+
+
+	@author Marcel P. | Plomi.net
+	@github https://github.com/plomi-net
+
+	@version 1.0.0
+
+]]
+
+local i2c, string = i2c, string
+local issetBit = bit.isset
+local setBit = bit.set
+local clearBit = bit.clear
+
+local MCP23017_ADDRESS = 0x20 -- (0x20 to 0x27)
+-- MCP23017_SDA = 2 -- D2/GPIO 4
+-- MCP23017_SCL = 1 -- D1/GPIO 5
+
+local OUTPUT = false
+local INPUT = true
+local HIGH = true
+local LOW = false
+
+-- registers (not used registers are commented out)
+local MCP23017_IODIRA = 0x00
+local MCP23017_IODIRB = 0x01
+local MCP23017_DEFVALA = 0x06
+local MCP23017_DEFVALB = 0x07
+local MCP23017_GPIOA = 0x12
+local MCP23017_GPIOB = 0x13
+
+--[[
+local MCP23017_IPOLA = 0x02
+local MCP23017_IPOLB = 0x03
+local MCP23017_GPINTENA = 0x04
+local MCP23017_GPINTENB = 0x05
+local MCP23017_DEFVALA = 0x06
+local MCP23017_DEFVALB = 0x07
+local MCP23017_INTCONA = 0x08
+local MCP23017_INTCONB = 0x09
+local MCP23017_IOCON = 0x0A
+local MCP23017_IOCON2 = 0x0B
+local MCP23017_GPPUA = 0x0C
+local MCP23017_GPPUB = 0x0D
+local MCP23017_INTFA = 0x0E
+local MCP23017_INTFB = 0x0F
+
+local MCP23017_INTCAPA = 0x10
+local MCP23017_INTCAPB = 0x11
+local MCP23017_OLATA = 0x14
+local MCP23017_OLATB = 0x15
+]]
+
+-- internal
+local MCP23017_id = 0
+local MCP23017_SPEED = i2c.SLOW
+local MCP23017_DEVICE_OK = false
+
+-- check device is available on address
+local function checkDevice(address)
+    i2c.start(MCP23017_id)
+    local response = i2c.address(MCP23017_id, address, i2c.TRANSMITTER)
+    i2c.stop(MCP23017_id)
+    if response ~= true then
+        print("MCP23017 device on " .. string.format('0x%02X', address) .. " not found")
+    end
+    return response
+end
+
+-- check device address (0x20 to 0x27)
+local function checkAddress(address)
+    local addr = tonumber(address)
+    if (addr > 31 and addr < 40) then
+        return true
+    else
+        print("MCP23017 address is out of range")
+        return false
+    end
+end
+
+-- write byte
+local function writeByte(registerAddr, val)
+    i2c.start(MCP23017_id)
+    i2c.address(MCP23017_id, MCP23017_ADDRESS, i2c.TRANSMITTER)
+    i2c.write(MCP23017_id, registerAddr)
+    i2c.write(MCP23017_id, val)
+    i2c.stop(MCP23017_id)
+end
+
+-- read byte
+local function readByte(registerAddr)
+    i2c.start(MCP23017_id)
+    i2c.address(MCP23017_id, MCP23017_ADDRESS, i2c.TRANSMITTER)
+    i2c.write(MCP23017_id, registerAddr)
+    i2c.stop(MCP23017_id)
+    i2c.start(MCP23017_id)
+    i2c.address(MCP23017_id, MCP23017_ADDRESS, i2c.RECEIVER)
+    local data = i2c.read(MCP23017_id, 1)
+    i2c.stop(MCP23017_id)
+    return string.byte(data)
+end
+
+-- get IO dir register
+local function getDirRegisterAddr(pin)
+    if pin > 7 then
```

Comment:
This function doesn't check `> 15` unlike `convertPin` below; is that OK?

Reply:
Many thanks for the advice. This is no longer necessary due to the changes to an a/b parameter.